From 4efafe3d5b10e23efdc4717c783f787d464ce25c Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 11:40:38 -0400
Subject: [net] ethtool: exit the loop when invalid index occurs

Message-id: <1410176446-10440-3-git-send-email-ivecera@redhat.com>
Patchwork-id: 91650
O-Subject: [RHEL7.1 PATCH 02/10] ethtool: exit the loop when invalid index occurs
Bugzilla: 1138823
RH-Acked-by: Michal Schmidt <mschmidt@redhat.com>
RH-Acked-by: Jarod Wilson <jarod@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>

BZ: #1138823

Upstream commit(s):
commit 266a16468432a77f2e1395b3de5df3ca699b1a9a
Author: Jean Sacren <sakiwit@gmail.com>
Date:   Sun Apr 27 02:20:38 2014 -0600

    ethtool: exit the loop when invalid index occurs

    The commit 3de0b592394d ("ethtool: Support for configurable RSS hash
    key") introduced a new function ethtool_copy_validate_indir() with
    full iteration of the loop to validate the ring indices, which could
    be an overkill. To minimize the impact, we ought to exit the loop as
    soon as the invalid index occurs for the very first time. The
    remaining loop simply doesn't serve any more purpose.

    Signed-off-by: Jean Sacren <sakiwit@gmail.com>
    Cc: Venkata Duvvuru <VenkatKumar.Duvvuru@Emulex.Com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/core/ethtool.c b/net/core/ethtool.c
index e376155..cc4f9fa 100644
--- a/net/core/ethtool.c
+++ b/net/core/ethtool.c
@@ -557,8 +557,10 @@ static int ethtool_copy_validate_indir(u32 *indir, void __user *useraddr,
 
  /* Validate ring indices */
  for (i = 0; i < size; i++) {
-  if (indir[i] >= rx_rings->data)
+  if (indir[i] >= rx_rings->data) {
    ret = -EINVAL;
+   break;
+  }
  }
  return ret;
 }
-- 
1.7.1